subsys/net/lib/nrf_cloud/src/nrf_cloud_cell_pos.c

```diff
@@ -0,0 +1,356 @@
+
+	cell_based_loc.lat = lat->valuedouble;
+	cell_based_loc.lon = lon->valuedouble;
+	cell_based_loc.unc = unc->valueint;
+	cell_based_loc.type = type;
+
+	LOG_DBG("Cell location: (%lf, %lf), unc: %d, type: %d",
+		cell_based_loc.lat, cell_based_loc.lon,
+		cell_based_loc.unc, cell_based_loc.type);
+
+	return 0;
+}
+
+static int parse_cell_location_response(const char *const buf)
+{
+	int ret;
+	cJSON *cell_loc_obj;
+	cJSON *data_obj;
+	enum cell_based_location_type cell_loc_type;
+
+	if (buf == NULL) {
+		return -EINVAL;
+	}
+
+	cell_loc_obj = cJSON_Parse(buf);
+	if (!cell_loc_obj) {
+		LOG_DBG("No JSON found for cell location");
+		return 1;
+	}
+
+	ret = -ENOTSUP;
+
+	/* Check for valid appId and msgType */
+	if (!json_item_string_exists(cell_loc_obj, CELL_POS_JSON_MSG_TYPE_KEY,
+				     CELL_POS_JSON_MSG_TYPE_VAL_DATA)) {
+		LOG_DBG("Wrong msg type for cell location");
+		goto cleanup;
+	}
+
+	if (json_item_string_exists(cell_loc_obj, CELL_POS_JSON_APPID_KEY,
+				    CELL_POS_JSON_APPID_VAL_SINGLE_CELL)) {
+		cell_loc_type = CELL_LOC_TYPE_SINGLE;
+	} else if (json_item_string_exists(cell_loc_obj, CELL_POS_JSON_APPID_KEY,
+					   CELL_POS_JSON_APPID_VAL_MULTI_CELL)) {
+		cell_loc_type = CELL_LOC_TYPE_MULTI;
+	} else {
+		LOG_DBG("Wrong app id for cell location");
+		goto cleanup;
+	}
+
+	data_obj = cJSON_GetObjectItem(cell_loc_obj, CELL_POS_JSON_DATA_KEY);
+	if (!data_obj) {
+		LOG_DBG("Data object not found in cell-based location msg.");
+		ret = -EBADMSG;
+		goto cleanup;
+	}
+
+	ret = json_parse_cell_location(data_obj, cell_loc_type);
+
+cleanup:
+	cJSON_Delete(cell_loc_obj);
+	return ret;
+}
+
+int nrf_cloud_request_cell_location(enum cell_based_location_type type,
+					 const bool request_loc)
+{
+	int err;
+	cJSON *data_obj;
+	cJSON *cell_pos_req_obj;
+
+	/* TODO: currently single cell only */
+	ARG_UNUSED(type);
+	cell_pos_req_obj = json_create_req_obj(CELL_POS_JSON_APPID_VAL_SINGLE_CELL,
+					   CELL_POS_JSON_MSG_TYPE_VAL_DATA);
+	data_obj = cJSON_AddObjectToObject(cell_pos_req_obj, CELL_POS_JSON_DATA_KEY);
+
+	if (!cell_pos_req_obj || !data_obj) {
+		err = -ENOMEM;
+		goto cleanup;
+	}
+
+	/* Add modem info to the data object */
+	err = json_add_modem_info(data_obj);
+	if (err) {
+		LOG_ERR("Failed to add modem info to cell loc req: %d", err);
+		goto cleanup;
+	}
+
+	/* By default, nRF Cloud will send the location to the device */
+	if (!request_loc) {
+		/* Specify that location should not be sent to the device */
+		cJSON_AddNumberToObject(data_obj,
+					CELL_POS_JSON_CELL_LOC_KEY_DOREPLY,
+					0);
+	}
+
+	err = json_send_to_cloud(cell_pos_req_obj);
+
+cleanup:
+	cJSON_Delete(cell_pos_req_obj);
+
+	return err;
+}
+
+int nrf_cloud_get_last_cell_location(double *const lat,
+					  double *const lon)
+{
+	if (!lat && !lon) {
+		return -EINVAL;
+	}
+
+	if (lat) {
+		*lat = cell_based_loc.lat;
+	}
+
+	if (lon) {
+		*lon = cell_based_loc.lon;
+	}
+
+	return 0;
+}
+
+int nrf_cloud_cell_pos_process(const char *buf, size_t buf_len)
```

Comment:
i think it would be useful to add a parameter to this function, which can hold the result.
The following struct (from an ongoing PR) https://github.com/nrfconnect/sdk-nrf/blob/217876573b6734d736b49444cf05d9ddba53f633/include/net/nrf_cloud.h#L239 could be added to nrf_cloud.h now.  

This function could then be updated as follows:

`int nrf_cloud_cell_pos_process(const char *buf, size_t buf_len, struct nrf_cloud_cell_location_result *const result)`

The result pointer could be NULL.  If the user wants the data, they would provide a valid pointer.  
This would eliminate the need for the user to call `nrf_cloud_get_last_cell_location`.
It would also give the user the type and uncertainty.

Reply:
Yes we can, but should the data structure defined in nrf_cloud_cell_pos.h instead?

Reply:
Shall we add this in next PR, after your on-going PR or even better after MCELL is available?

Reply:
since this changes the API, i would recommend just adding this parameter now.  
It eliminates the need call to `nrf_cloud_get_last_cell_location`. 
in fact, `nrf_cloud_get_last_cell_location` could probably be removed and the cell_pos lib would not need to store the last value.

Reply:
OK, I'll then make some major changes of below:
.remove `nrf_cloud_get_last_cell_location` and no static location data storage in nRF Cloud lib
.move the location data structure to API header
.change `nrf_cloud_cell_pos_process` to return the parsed location info instantly
.apply all the naming changes your suggested below